admin/server/auth/handlers.go

```diff
@@ -37,6 +38,44 @@ func (a *Authenticator) RegisterEndpoints(mux *http.ServeMux) {
 	mux.Handle("/auth/", observability.Middleware("admin", a.logger, inner))
 }
 
+func (a *Authenticator) authWithToken(w http.ResponseWriter, r *http.Request) {
+	// Generate random state for CSRF
+	b := make([]byte, 32)
+	_, err := rand.Read(b)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to generate state: %s", err), http.StatusInternalServerError)
+		return
+	}
+	state := base64.StdEncoding.EncodeToString(b)
+
+	// Get auth cookie
+	sess := a.cookies.Get(r, cookieName)
+
+	// Set state in cookie
+	sess.Values[cookieFieldState] = state
+
+	// Get redirect destination
+	redirect, ok := sess.Values[cookieFieldRedirect].(string)
+	if !ok || redirect == "" {
+		redirect = "/"
+	}
+
+	// Set auth token in cookie
+	token := r.URL.Query().Get("token")
+	if token != "" {
+		sess.Values[cookieFieldAccessToken] = token
+	}
+
+	// Save cookie
+	if err := sess.Save(r, w); err != nil {
+		http.Error(w, fmt.Sprintf("failed to save session: %s", err), http.StatusInternalServerError)
+		return
+	}
+
+	// Redirect to UI (usually)
+	http.Redirect(w, r, redirect, http.StatusTemporaryRedirect)
```

Comment:
This should be:

```go
	// Get new auth token
	newToken := r.URL.Query().Get("token")
	if newToken == "" {
		http.Error(w, "token not provided", http.StatusBadRequest)
		return
	}

	// Get auth cookie
	sess := a.cookies.Get(r, cookieName)

	// If there's already a token in the cookie, revoke it (since we're now setting a new one)
	oldAuthToken, ok := sess.Values[cookieFieldAccessToken].(string)
	if ok && oldAuthToken != "" {
		err := a.admin.RevokeAuthToken(r.Context(), oldAuthToken)
		if err != nil {
			a.logger.Error("failed to revoke old user auth token during new auth", zap.Error(err), observability.ZapCtx(r.Context()))
			// The old token was probably manually revoked. We can still continue.
		}
	}

	// Set auth token in cookie
	sess.Values[cookieFieldAccessToken] = newToken

	// Save cookie
	if err := sess.Save(r, w); err != nil {
		http.Error(w, err.Error(), http.StatusInternalServerError)
		return
	}

	// Redirect to UI
	http.Redirect(w, r, a.opts.FrontendURL, http.StatusTemporaryRedirect)
```